Add a section about avoiding XSS in JavaScript

dkoo opened the pull request:
Hat tip to @smy315 for the initial writeup and solution examples. Given the recent conversation around this (and running into it in various projects), I thought XSS warranted a section in the JavaScript best practices.

Commit:
Add a section to the JavaScript page about XSS and best practices for avoiding it

Commit:
Fix a syntax error in one of the XSS examples

JodiWarren commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
 
+### Secure Your Code
+
+In JavaScript, we often have to insert new elements with dynamic attributes and content into the DOM. A common way to do this is to use the ```innerHTML``` method like so:
````
Maybe you could provide link to the MDN page here https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML

JodiWarren commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+#### Use ```textContent``` instead of ```innerHTML```
+
+When setting the human-readable content of a single element, using ```textContent``` is safer than using ```innerHTML``` because it does not parse strings as HTML—meaning any malicious code passed to it will not be executed.
````
Maybe make ```textContent``` link to the MDN page: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent

smy315 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+```javascript
+const myElement = document.getElementById( 'myElement' );
+
+myElement.textContent = 'This will get displayed as human-readable text, even if it contains HTML-like <syntax>.'
````
Semicolon missing.

JodiWarren commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+```javascript
+// Using DOMPurify
+DOMPurify.sanitize( '<a href="#" onclick="alert( 'Alert!' );">link</a>' ); // Becomes <a href="#">link</a>
````
It might be nice to provide the same example as on line 163 here, just to keep the examples consistent.

dkoo replied:
Nice idea—I added it as the first example for both DOMPurify and js-xss. Also updated the output for js-xss to reflect what it actually was upon testing.

smy315 commented on _includes/markdown/JavaScript.md:
```diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+// Set attributes and content.
+container.setAttribute( 'class', 'container' );
+anchor.setAttribute( 'href', someUrl );
```
Put someUrl in single quotes or assign a url to this variable.

dkoo replied:
I assigned it a value in all examples.

smy315 commented on _includes/markdown/JavaScript.md:
```diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+// Set attributes and content.
+container.setAttribute( 'class', 'container' );
+anchor.setAttribute( 'href', someUrl );
+anchor.textContent = someContent;
```
Put someContent in single quotes.

dkoo replied:
I assigned it a value instead in all examples.

smy315 reviewed (changes requested):
Looks good, left few minor comments on the examples.

JodiWarren commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+In JavaScript, we often have to insert new elements with dynamic attributes and content into the DOM. A common way to do this is to use the ```innerHTML``` method like so:
+
+```javascript
+const myElement = document.getElementById( 'myElement' );
````
Maybe change `myElement` to `element` to differentiate it from the elements that you're creating. It's has a pretty mild effect here, but in the more verbose `textContent` example it might add a little bit of clarity.

JodiWarren reviewed (approved):
Nice work dude! I made a few suggestions - but please feel free to take or ignore them!

smy315 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,74 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+#### Use the ```Element``` API instead of HTML-parsing methods
+
+When you need to create multiple DOM elements, use JavaScript's [```Element``` API](https://developer.mozilla.org/en-US/docs/Web/API/Element) to create elements, set attributes and append them to the document. Creating your own elements and attributes will ensure that only those you explicitly define will make their way into the DOM. Refactoring our first example:
````
Extra space linked before API.

nicholasio reviewed (commented):
This looks great to me. I'd add a note saying that including the library is only worth it if the code base is generating a lot of HTML dynamically. For some projects we have just a few spots whee JS is generating HTML that is generally better to just use the Element API IMO.

Commit:
Updates to address code review feedback

dkoo commented:
@JodiWarren Thanks for the suggestions! I took most of them.

dkoo commented:
@nicholasio I added a note in the beginning of the last subsection stating that the `Element` API is the preferred best practice, but that a sanitizer is another option for projects that need to generate a lot of dynamic HTML.

dkoo commented:
@smy315 Feedback addressed!

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+someElement.innerHTML = `<div class="container"><a href="${ someUrl }">${ someContent }</a></div>`;
+```
+
+However, passing HTML strings to ```innerHTML``` and methods like it (including jQuery's commonly used ```append``` and ```html``` methods) can expose your code to cross-site scripting, also known as XSS—the most common security vulnerability in JavaScript. Because these methods evaluate strings passed to them as HTML, they can execute potentially harmful code. For example, what happens if ```som ...
````
@dkoo In this paragraph:

no need to reference jQuery as we're migrating away from it.
space between XSS and "the"?
remove "what happens if"

dkoo replied:
Removed the jQuery reference and reworded the "what happens if". No space after "XSS"—using an em dash there!

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+When setting the human-readable content of a single element, using [```textContent```](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) is safer than using ```innerHTML``` because it does not parse strings as HTML—meaning any malicious code passed to it will not be executed.
+
+```javascript
````
@dkoo can we replace this code block with a link to documentation about using `textContent`?

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+When you need to create multiple DOM elements, use JavaScript's [```Element```](https://developer.mozilla.org/en-US/docs/Web/API/Element) API to create elements, set attributes and append them to the document. Creating your own elements and attributes will ensure that only those you explicitly define will make their way into the DOM. Refactoring our first example:
+
+```javascript
````
@dkoo there's a ton of stuff in the Element API, rather than addressing a very small portion of it, let's just link off to MDN.

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+#### Sanitize HTML strings before appending to the DOM
+
+In general, using the ```Element``` API is the preferred best practice to safely create and add DOM elements. However, you'll notice that the refactored example is much more verbose than the original example—what was just a few lines of executable code more than doubled after refactoring. This can become painful if you need to create many new elements. In those cases, the convenience of methods l ...
````
@dkoo reword this a little if the code block above it gone

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+If you need to generate a large amount of HTML dynamically, consider adding a client-side sanitization library to your project so you can strip potentially malicious code from your HTML before you add it to the DOM. Two popular libraries are [DOMPurify](https://github.com/cure53/DOMPurify) and [js-xss](https://github.com/leizongmin/js-xss). They work similarly:
+
+```javascript
````
@dkoo If we want to sanction this library, that's fine with me, but I don't really want to have to maintain third party code blocks in our best practice documentation. Let's remove this block in favor of a link to documentation.

timwright12 commented on _includes/markdown/JavaScript.md:
```diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+DOMPurify.sanitize( '<svg><g/<script></svg>' ); // Becomes <svg><g></g></svg>
+
+// Using js-xss
+filterXSS( `<div class="container"><a href="https://someurl.com/"><img src="fakeImage" onerror="alert( 'hacked' )" /></a></div>` ); // Becomes <div><a href="https://someurl.com/"><img src /></a></div>
```
@dkoo same comment about third party code

timwright12 commented on _includes/markdown/JavaScript.md:
```diff
@@ -155,6 +155,80 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+Note that appending new elements to the DOM is a relatively expensive operation, so generally you'll want to build the structure of new, nested elements _before_ adding them to the DOM, then append them to an existing element all at once by appending a single container element.
+
+#### Sanitize HTML strings before appending to the DOM
```
@dkoo is it worth mentioning DOMParser here?

Commit:
Updates to address further code review feedback

dkoo commented:
@timwright12 Thanks for the feedback! I replaced most of the code blocks with links to the MDN documentation and removed references to the specific 3rd party libraries. Also added a paragraph about `DOMParser`, as you suggested.

timwright12 commented on _includes/markdown/JavaScript.md:
````diff
@@ -155,6 +155,43 @@ On all new projects you should be using up to date JavaScript methodologies comb
+
+When setting the human-readable content of a single element, using ```textContent``` is safer than using ```innerHTML``` because it does not parse strings as HTML—meaning any malicious code passed to it will not be executed. Refer to [MDN's documentation on ```textContent```](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) for more info.
+
+```
````
@dkoo rogue dashes?

jonathantneal replied:
You’ll need to remove this rogue line. It’s causing formatting issues.

Commit:
Kill rogue code block fence

dkoo commented:
@timwright12 @jonathantneal Whoops, rogue characters deleted!

jonathantneal commented:
I’m a _big_ fan of this new “Secure Your Code” section. Well done, @dkoo!
